File: num_0s_between_1s.c
/*
 * Juntong Liu
 *            2023.02.22: 18:12
 *  
 * File name: num_0s_between_1s.c
 *
 * This program check a input integer and counts those continuous 0 bits between two 1 bits in the integer, 
 * save all lengthes in an array and print out the length of the longest number of 0 bits between two 1 bits.
 * For example, 
 * 
 * If the integer is 5678 = 1011000101110          The longest length of continuous 0 bits between two 1 bits is: 3
 *                              ^^^
 *
 * If the integer is 123  = 1111011                The longest length of continuous 0 bits between two 1 bits is: 1
 *                              ^
 * 
 * If the integer is 8765 = 10001000111101         The longest length of continuous 0 bits between two 1 bits is: 3
 *         
 * To compile and run the program:
 *       
 *  $ gcc -Wall num_0s_between_1s.c -o num_0s_between_1s
 *  $ ./num_0s_between_1s
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <limits.h>

static void print_binary(unsigned int integer)
{
	int i = CHAR_BIT * sizeof integer; /* bits in an integer */
	while (i--)
	{
		putchar('0' + ((integer >> i) & 1));
	}
}

static int mycompare( const void* a, const void* b)
{
   int int_a = * ( (int*) a );
   int int_b = * ( (int*) b );

   // an easy expression for comparing
   return (int_a > int_b) - (int_a < int_b);
}

/* find the length of all continuous 0 bits between two 1 bits and store those length in
 * an array. Return the longest length to the caller.
 */
static int num_z(int n)
{
#define INTEGER_SIZE CHAR_BIT * sizeof(int)
	int oneb = 1;
	int integer = n;
	int count = 0;
	int resultindex = 0;
	int checkedb = 0;
	int result[INTEGER_SIZE/2];
	
	while(checkedb < INTEGER_SIZE)
	{
		if((oneb & integer) == 0)
		{
			oneb <<= 1;
			checkedb++;
		}
		else
			break;
	}

	while(checkedb < INTEGER_SIZE) 
	{
		while(checkedb < INTEGER_SIZE)
		{
			if((oneb & integer) != 0)
			{
				checkedb++;
				oneb <<= 1;	
			}
			else
				break;
		}

		while(checkedb < INTEGER_SIZE)
		{
			if((oneb & integer) == 0)         
			{
				count++;
				checkedb++;
				oneb <<= 1;
			}
			else{
				result[resultindex] = count;
				resultindex++;
				count = 0;
				break;
			}
		}
	}
  
	qsort(result, resultindex, sizeof(int), mycompare);
	
	if(resultindex == 0)
		return 0;            // no 0 bits between two 1s in the integer
	else{
		for(int i=0; i<resultindex; i++)
			printf("%d. has %d zero\n", i, result[i]);
		printf("The shortest length of continuous 0 bits between two 1 bits is: %d\n", result[0]);
		resultindex--;
		return (result[resultindex]);
	}
}

int main()
{
	char buff[8];
	int your_number;
	
	while(1)
	{
		
		printf("Type in a integer number, or type in q, Q to quit.\n");
		fgets(buff, sizeof(buff), stdin);
		if((buff[0] == 'q') || (buff[0] == 'Q'))
			return 0;

		your_number = atoi(buff);
		if(your_number == 0)
		{
			printf("Your input is 0 or no-digital! Try again!\n");
			continue;
		}
		
		printf("You have typed in: %d = ", your_number);
		print_binary((unsigned int)your_number);
		printf("\n");
		int result = num_z(your_number);

		printf("The max continuous 0 bits between two 1 bits = %d\n", result);
	}
	return 0;
}

File: num_0s_between_1s.rs
/*
 * Juntong Liu
 *            2023.03.07: 18:12
 *  
 * File name: num_0s_between_1s.rs
 *
 * This program takes an integer from user and counts those continuous 0 bits between two 1 bits in the integer. 
 * Those lengthes of continuous 0 bits between two 1 bits will be saved in an arry and the longest length will be printed out.
 * 
 * For example, on a machine where integer_size = 32 bits, 
 * 
 *  5678 = 00000000000000000001011000101110      The longest length of continuous 0 bits between two 1 bits is: 3
 *                                ^^^
 *
 *   123 = 00000000000000000000000001111011      The longest length of continuous 0 bits between two 1 bits is: 1
 *                                      ^
 * 
 *  -123 = 11111111111111111111111110000101      The longest length of continuous 0 bits between two 1 bits is: 4
 *                                  ^^^^
 * To compile and run the program:
 *       
 *  $ rustc num_0s_between_1s.rs
 *  $ ./num_0s_between_1s
 *
 */

fn num_z(n: i32) {
    let mut oneb: i32 = 1;
    let mut checkedb = 0;
    const INT_SIZE: i32 = 32;
    let mut count: usize = 0; 
    let mut maxlong: Vec<usize> = Vec::new(); 
    let integer: i32 = n;
    
    while checkedb < INT_SIZE {   
        if oneb & integer == 0 {
            oneb <<= 1;
            println!("1: oneb is: {}", oneb);
            checkedb += 1;
        }
        else {
            break;
        }
    }

    while checkedb < INT_SIZE {          
        while checkedb < INT_SIZE {       
            if oneb & integer != 0 {      
                checkedb += 1;
                oneb <<= 1;
            }
            else {
                break;
            }
        }
        while checkedb < INT_SIZE {
            if oneb & integer == 0 {   
                count += 1;
                checkedb += 1;
                oneb <<= 1;
            }
            else { 
                  maxlong.push(count);
                  count = 0;
                break;
            }
        }
    }

    if maxlong.len() > 0 {
        maxlong.sort();
        let ln = maxlong.pop();
        match ln {
            Some(x) => println!("The maximum length of continous 0s between two 1s is: {}", x),
            None => (), 
        }
    }
}
  
fn main() 
{
    let mut user_in = String::new();
    
    loop{
        user_in.clear();
        println!("Please enter an integer number or enter q or Q to quit the program:");
        std::io::stdin()
            .read_line(&mut user_in)
            .expect("Failed to read in line!");
        
        let user_in = user_in.trim();
        if user_in == "q" || user_in == "Q" {
            return; 
        }

        let a_integer: i32 = match user_in.trim().parse()
        {   
            Ok(num) => num,
            Err(_)  => {
                println!("Your input is not a number, try again!");
                continue;
            }
        };
        println!("You have entered: {:b}", a_integer);
        num_z(a_integer); 
    }
}
